src/foundation/dictionary.rs

```diff
@@ -38,9 +46,103 @@ impl NSMutableDictionary {
         }
     }
 
-    /// Consumes and returns the underlying `NSMutableDictionary`.
-    pub fn into_inner(mut self) -> id {
-        &mut *self.0
+    /// Returns the value associated with a given key. Corresponds to objectForKey:
+    pub fn get(&self, key: &str) -> Option<id> {
+        let key = NSString::new(key);
+        let id: id = unsafe { msg_send![self.0, objectForKey: key] };
+
+        if !id.is_null() {
+            Some(id)
+        } else {
+            None
+        }
+    }
+
+    /// A new array containing the dictionary’s keys, or an empty array if the dictionary has no entries. Corresponds to allKeys
+    ///
+    /// **NOTE:** This only works with string keys
+    pub fn keys(&self) -> Vec<String> {
+        let keys = NSArray::retain(unsafe { msg_send![self.0, allKeys] });
+        keys.iter().map(|s| NSString::retain(s).to_string()).collect()
+    }
+
+    /// Converts the dictionary into a hashmap, passing each item through a transform function.
+    ///
+    /// **NOTE:** This only works with string keys
+    pub fn into_hashmap<T, F>(&self, item_transform: F) -> HashMap<String, T>
+    where
+        F: Fn(&String, id) -> T,
+    {
+        let mut map = HashMap::new();
+
+        let keys = self.keys();
+
+        for key in keys {
+            let item_id = self.get(&key);
+
+            if let Some(item_id) = item_id {
+                let item = item_transform(&key, item_id);
+
+                map.insert(key, item);
+            } else {
+                // TODO: Should there be an assertion here for runtime failure?
+                continue;
+            }
+        }
+
+        map
+    }
```

Comment:
I hate these being string specific, but I'm not sure how else to keep this API usable

Reply:
This is fine to use String keys for now, to be honest - better to be out and working with potential improvements/revamps later, especially since while I'm fine with these pieces being more open, they're still mostly an internal aspect.

I think you're probably on the right train of thought re: having an assertion here.